codegen/ace/src/circuit.rs

```diff
@@ -0,0 +1,507 @@
+                Operation::Add(l, r) => {
+                    res.insert(out, res[l] + res[r]);
+                }
+                Operation::Sub(l, r) => {
+                    res.insert(out, res[l] - res[r]);
+                }
+                Operation::Mul(l, r) => {
+                    res.insert(out, res[l] * res[r]);
+                }
+            }
+        }
+        res[&root]
+    }
+}
+
+impl std::ops::Index<NodeIndex> for Circuit {
+    type Output = Operation;
+    fn index(&self, i: NodeIndex) -> &Operation {
+        let i: usize = NodeIndex::into(i);
+        &self.0[i]
+    }
+}
+
+/// Iterate over the nodes in the DAG depth-first.
+/// Properties of the result:
+/// - only dependencies of the given roots are returned
+/// - parents come after their children
+/// - no duplicates (e.g. when two contraints share a subexpression)
+/// - inputs are preserved even if unused
+pub struct UniqueOperationIterator<'a> {
+    // circuit being iterated on
+    circuit: &'a Circuit,
+    // nodes visited but not necessarily processed yet
+    visited: BTreeSet<NodeIndex>,
+    // nodes already processed, meaning their children are visited
+    processed: BTreeSet<NodeIndex>,
+    // nodes to be processed
+    process: Vec<NodeIndex>,
+}
+impl<'a> UniqueOperationIterator<'a> {
+    pub fn new(circuit: &'a Circuit, roots: Vec<NodeIndex>) -> Self {
+        let visited = BTreeSet::new();
+        let processed = BTreeSet::new();
+        let process = roots.into_iter().chain(circuit.get_inputs()).collect();
+        UniqueOperationIterator {
+            circuit,
+            visited,
+            processed,
+            process,
+        }
+    }
+}
+impl Iterator for UniqueOperationIterator<'_> {
+    type Item = (NodeIndex, Operation);
+    fn next(&mut self) -> Option<Self::Item> {
+        use Operation::*;
+
+        while let Some(id) = self.process.pop() {
+            let op = self.circuit[id];
+            if self.processed.contains(&id) {
+                return Some((id, op));
+            } else {
+                self.process.push(id);
+                self.visited.insert(id);
+                self.processed.insert(id);
+                match op {
+                    Value(_) => {}
+                    Add(l, r) | Sub(l, r) | Mul(l, r) => {
+                        if !self.visited.contains(&r) {
+                            self.visited.insert(r);
+                            self.process.push(r)
+                        };
+                        if !self.visited.contains(&l) {
+                            self.visited.insert(l);
+                            self.process.push(l)
+                        };
+                    }
+                };
+            }
+        }
+        None
+    }
+}
+
+/// Remap the children of an operation. Typically when migrating an operation
+/// between data structures.
+pub fn remap_op(op: Operation, mapping: &BTreeMap<NodeIndex, NodeIndex>) -> Operation {
+    use Operation::*;
```

Comment:
Importing all enum variants is generally discourages since it may lead to incorrect matches when the enum variants change. In this case it would be okay, but if the variants did not have fields inside them, a removed field would be matched as the "catch-all" case.